proof_of_concept/registry.py

```diff
@@ -5,194 +5,279 @@
-        """Returns the public key of the given party.
 
-        Args:
-            party: Name of the party to look up.
+class PolicyServerDescription(RegisteredObject):
+    """Describes a policy server to the rest of the DDM.
 
-        Return:
-            The public key of the given party.
+    Attributes:
+        namespace: The namespace governed by this server.
+        site: The site at which this server is located.
 
-        Raises:
-            KeyError: If no party with the given name is registered.
-        """
-        return self._party_keys[party]
+    """
+    def __init__(
+            self, site: SiteDescription, namespace: NamespaceDescription,
+            server: IPolicyServer) -> None:
+        """Create a PolicyServerDescription.
 
-    def list_runners(self) -> List[str]:
-        """List names of all registered runners.
+        Args:
+            site: The site at which this server is located.
+            namespace: The namespace governed by this server.
+            server: The policy server.
 
-        Returns:
-            A list of names as strings.
         """
-        return list(self._runners.keys())
+        self.site = site
+        self.namespace = namespace
+        self.server = server
 
-    def get_runner(self, name: str) -> ILocalWorkflowRunner:
-        """Look up a LocalWorkflowRunner.
 
-        Args:
-            name: The name of the runner to look up.
+class Registry:
+    """Global registry of remote-accessible things.
 
-        Return:
-            The runner with that name.
+    Registers runners, stores, and assets. In a real system, runners
+    and stores would be identified by a URL, and use the DNS to
+    resolve. For now the registry helps with this.
+    """
+    def __init__(self) -> None:
+        """Create a new registry."""
+        self._assets = dict()           # type: Dict[str, str]
 
-        Raises:
-            KeyError: If no runner with the given name is
-                    registered.
-        """
-        return self._runners[name]
+        self._archive = ReplicableArchive[RegisteredObject]()
+        self._store = CanonicalStore[RegisteredObject](self._archive)
+        self.replication_server = ReplicationServer[RegisteredObject](
+                self._archive, 1.0)
 
-    def get_runner_admin(self, name: str) -> str:
-        """Look up who administrates a given runner.
+    def register_party(
+            self, name: str, namespace: str, public_key: RSAPublicKey) -> None:
+        """Register a party with the DDM.
 
         Args:
-            name: The name of the runner to look up.
+            name: Name of the party.
+            namespace: ID namespace owned by this party.
+            public_key: Public key of this party.
+        """
+        for i in self._store.objects():
+            if isinstance(i, PartyDescription) and i.name == name:
+                raise RuntimeError('There is already a party called {}'.format(
```

Comment:
I would raise ValueError here

Reply:
Good point!

Reply:
Sorry, I've changed my mind :-). The Python docs say that `ValueError` should be raised if an "inappropriate" value is passed to a function. There are a few subclasses which define errors that are raised when one tries to process a bytes string as Unicode, but it doesn't form a valid UTF-8 sequence (or so I imagine :-)). So in that case, the contents of the variable are invalid.

Here, if you try to register a party twice, the name of the party is not invalid (presumably it was fine when it was first registered). It's just that the system is in a state in which the requested operation is impossible. So unless we define a subclass of RuntimeError for this specifically, I think RuntimeError is actually better.

Reply:
I would still raise ValueError here (but maybe that is because people often misuse it). But I  am totally fine with RuntimeError! :)